src/consumer/consumerGroup.js

```diff
@@ -406,162 +435,110 @@ module.exports = class ConsumerGroup {
         }
-      }
-
-      const requests = keys(requestsPerNode).map(async nodeId => {
-        const broker = await this.cluster.findBroker({ nodeId })
-        const { responses } = await broker.fetch({
-          maxWaitTime,
-          minBytes,
-          maxBytes,
-          isolationLevel: this.isolationLevel,
-          topics: requestsPerNode[nodeId],
-          rackId: this.rackId,
-        })
-
-        const batchesPerPartition = responses.map(({ topicName, partitions }) => {
-          const topicRequestData = requestsPerNode[nodeId].find(({ topic }) => topic === topicName)
-          let preferredReadReplicas = this.preferredReadReplicasPerTopicPartition[topicName]
-          if (!preferredReadReplicas) {
-            this.preferredReadReplicasPerTopicPartition[topicName] = preferredReadReplicas = {}
-          }
 
-          return partitions
-            .filter(
-              partitionData =>
-                !this.seekOffset.has(topicName, partitionData.partition) &&
-                !this.subscriptionState.isPaused(topicName, partitionData.partition)
-            )
-            .map(partitionData => {
-              const { partition, preferredReadReplica } = partitionData
-              if (preferredReadReplica != null && preferredReadReplica !== -1) {
-                const { nodeId: currentPreferredReadReplica } =
-                  preferredReadReplicas[partition] || {}
-                if (currentPreferredReadReplica !== preferredReadReplica) {
-                  this.logger.info(`Preferred read replica is now ${preferredReadReplica}`, {
-                    groupId: this.groupId,
-                    memberId: this.memberId,
-                    topic: topicName,
-                    partition,
-                  })
-                }
-                preferredReadReplicas[partition] = {
-                  nodeId: preferredReadReplica,
-                  expireAt: Date.now() + this.metadataMaxAge,
-                }
+        return partitions
+          .filter(
+            ({ partition }) =>
+              !this.seekOffset.has(topicName, partition) &&
+              !this.subscriptionState.isPaused(topicName, partition)
+          )
+          .map(partitionData => {
+            const { partition, preferredReadReplica } = partitionData
+
+            if (preferredReadReplica != null && preferredReadReplica !== -1) {
+              const { nodeId: currentPreferredReadReplica } = preferredReadReplicas[partition] || {}
+              if (currentPreferredReadReplica !== preferredReadReplica) {
+                this.logger.info(`Preferred read replica is now ${preferredReadReplica}`, {
+                  groupId: this.groupId,
+                  memberId: this.memberId,
+                  topic: topicName,
+                  partition,
+                })
               }
+              preferredReadReplicas[partition] = {
+                nodeId: preferredReadReplica,
```

Comment:
Once we have assigned the partition to a different broker, isn't it the case that a fetch for that broker could now execute before the batch from this broker has been processed and resolved?

Reply:
I will focus on this at the end of the week. 
On the first look, you might be right, though I'm not sure how often this could trigger, if at all, without rebalance.

Reply:
Just to see that I'm on the same page - the scenario you're describing is one where we have partition 1 assigned to broker 1. We fetch a batch from broker 1 and start to process it. While that's being processed, a Kafka admin goes and reassigns partition 1 to broker 2. We fetch from broker 2, and now we're basically processing the same messages in parallel, potentially causing some really strange behavior. Correct?

This does sound plausible. I haven't tried it to see what would happen, but I also can't find any information about how partition reassignment would affect consumer groups (other than if the leader changes), so I'm guessing it's something that needs to be handled client-side.

If so, then it seems a reasonable assertion to make that only one worker can process batches for a particular partition at any given time. Not sure where to enforce that though - if it would be in the worker by making the workers have to "check out" the partition they want to process messages for (then how do we deal with the already processed messages when the second worker picks them up from the second batch?), or if it can be enforced higher up somehow.

Reply:
Took a stab at improving this in https://github.com/tulios/kafkajs/commit/1c4b02e1dc477c380e5a2e4705f05491f14096ad I can't figure out how to create a PR towards your branch (I guess it's unusual to have the upstream create a PR towards a fork...), but you can see the diff and do a manual merge if you think it makes sense.

I still don't know if this is the right approach, or if we should do something more like the rebalancing when we detect this happening. Even with this, there's still no guarantee that worker 1 won't finish processing a batch for partition 1 just before worker 2 starts to process the same batch from partition 1, in which case they'll still be processing the same messages. The only way I see to avoid that is to look at the resolved offset and skip offsets before the currently resolved one.

Reply:
Btw, I asked the data streaming operations team at work if they knew what would happen in this scenario. They tested it out and verified that reassigning a partition to a different broker does not cause a consumer group rebalance, so it is something we have to handle on the client side.

Reply:
And for those reading along at home, the solution we went with in https://github.com/priitkaard/kafkajs/pull/2 was to have the FetchManager keep track of which fetcher is assigned to a particular topic-partition at any given time. When a fetcher receives a response, it filters out batches for any topic-partitions it is not currently assigned to. In the scenario we've discussed, this would mean that the second fetcher (which is now fetching from the newly assigned broker) would filter out any topic-partitions that the fetcher processing batches from the previously assigned broker is currently processing or has queued up.

A more elegant solution might be to avoid fetching the topic-partition at all, if it's already assigned to a different fetcher, instead of filtering out the results. But this scenario should be quite rare, so maybe the juice is not worth the squeeze.